db/db_impl.cc

```diff
@@ -1153,6 +1154,28 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
       }
       WriteBatchInternal::SetContents(&batch, record);
 
+#ifndef ROCKSDB_LITE
+      if (db_options_.wal_filter != nullptr) {
+        WALFilter::WALProcessingOption walProcessingOption =
+          db_options_.wal_filter->LogRecord(batch);
+
+        switch (walProcessingOption) {
+        case  WALFilter::WALProcessingOption::kContinueProcessing:
+          //do nothing, proceeed normally
+          break;
+        case WALFilter::WALProcessingOption::kIgnoreCurrentRecord:
+          //skip current record
+          continue;
+        case WALFilter::WALProcessingOption::kStopReplay:
+          //skip current record and stop replay
+          continue_replay_log = false;
+          continue;
+        default:
+          assert(false); //unhandled case
+        }
+      }
```

Comment:
From the source codes, it will be multiple LogData in one batch.

What you can do is to have the callback function to have an optional write batch returned. If it is not empty, we insert the returned write batch, instead of the original one. Would that work for you?